pyo3-macros-backend/src/utils.rs

```diff
@@ -67,14 +68,58 @@ pub fn option_type_argument(ty: &syn::Type) -> Option<&syn::Type> {
     None
 }
 
+// TODO: Replace usage of this by [`syn::LitCStr`] when on MSRV 1.77
+#[derive(Clone)]
+pub struct LitCStr {
+    lit: CString,
+    span: Span,
+    #[allow(dead_code)]
+    ctx: Ctx,
+}
+
+impl LitCStr {
+    pub fn new(lit: CString, span: Span, ctx: Ctx) -> Self {
+        Self { lit, span, ctx }
+    }
+
+    pub fn empty(ctx: Ctx) -> Self {
+        Self {
+            lit: CString::new("").unwrap(),
+            span: Span::call_site(),
+            ctx,
+        }
+    }
+}
+
+impl quote::ToTokens for LitCStr {
+    #[cfg(c_str_lit)]
+    fn to_tokens(&self, tokens: &mut TokenStream) {
+        syn::LitCStr::new(&self.lit, self.span).to_tokens(tokens);
+    }
+
+    #[cfg(not(c_str_lit))]
+    fn to_tokens(&self, tokens: &mut TokenStream) {
+        let Ctx { pyo3_path, .. } = &self.ctx;
+        let lit = self.lit.to_str().unwrap();
+        tokens.extend(quote::quote_spanned!(self.span => #pyo3_path::ffi::c_str!(#lit)));
+    }
+}
+
 /// A syntax tree which evaluates to a nul-terminated docstring for Python.
 ///
 /// Typically the tokens will just be that string, but if the original docs included macro
 /// expressions then the tokens will be a concat!("...", "\n", "\0") expression of the strings and
-/// macro parts.
-/// contents such as parse the string contents.
+/// macro parts. contents such as parse the string contents.
+#[derive(Clone)]
+pub struct PythonDoc(PythonDocKind);
+
 #[derive(Clone)]
-pub struct PythonDoc(TokenStream);
+enum PythonDocKind {
+    LitCStr(LitCStr),
+    // There is currently no way to `concat!` c-string literals, we fallback to the `c_str!` macro in
+    // this case.
+    Tokens(TokenStream),
```

Comment:
Agreed I think this case will have to stick around for the foreseeable future, but I think we would at least eventually be able to just use something like `CStr::from_bytes_with_nul` (which is const since 1.72).

Reply:
... or maybe #4270 is good enough for now :)